Consolidate bounding box code

harrism opened the pull request:
Fixes #746. Replaces the header-only `trajectory_bounding_boxes` with `point_bounding_boxes` and reuses it to implement `linestring_bounding_boxes` and `polygon_bounding_boxes`.

~By providing a single header-only function `point_bounding_boxes` that forms the implementation of `trajectory_bounding_boxes`, `linestring_bounding_boxes`, and `polygon_bounding_boxes`, this PR also closes closes 567 and closes 568.~

It does not, as `point_bounding_boxes` takes IDs, not polygon and ring offsets as the other APIs do. So we can't quite mark those as fixed.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

robertmaynard reviewed (commented):
I believe that `cpp/tests/CMakeLists.txt` needs to be updated since we renamed `experimental/trajectory/trajectory_bounding_boxes_test.cu`

harrism commented on cpp/tests/CMakeLists.txt:
```diff
@@ -143,8 +143,8 @@ ConfigureTest(POINT_IN_POLYGON_TEST_EXP
 
-ConfigureTest(TRAJECTORY_BOUNDING_BOXES_TEST_EXP
-    experimental/trajectory/trajectory_bounding_boxes_test.cu)
+ConfigureTest(POINT_BOUNDING_BOXES_TEST_EXP
```
@robertmaynard see here, I did modify this test.

harrism commented:
> I believe that `cpp/tests/CMakeLists.txt` needs to be updated since we renamed `experimental/trajectory/trajectory_bounding_boxes_test.cu` 

You must have missed where I updated it? It's now called POINT_BOUNDING_BOXES_TEST_EXP

isVoid reviewed (approved):
Good cleanup!

harrism commented on cpp/include/cuspatial/experimental/detail/bounding_box.cuh:
```diff
@@ -46,27 +60,32 @@ struct box_minmax {
+                                         IdInputIt ids_last,
+                                         PointInputIt points_first,
+                                         BoundingBoxOutputIt bounding_boxes_first,
+                                         T expansion_radius,
```
Discussed offline with @isVoid. Doing this would require doing the addition of the expansion radius to each bounding box after the reduction. We looked at a few of approaches.

1. `thrust::transform` after the `reduce_by_key`. This incurs an additional kernel launch. The cost of this may be larger than the cost of a couple of adds to each input coordinate, since the reduction should be bandwidth bound (so the adds are "free").
2. Use a `transform_output_iterator`. This doesn't work because the output type of a `reduce_by_key` bust be constructable from an instance of the input type. 
3. Leave the code as is (expand every point into a box using the expansion radius).

We chose option 3.

harrism commented on cpp/include/cuspatial/experimental/bounding_box.cuh:
```diff
@@ -16,22 +16,29 @@
+ * @brief Compute the spatial bounding boxes of sequences of points.
+ *
+ * Computes a bounding box around all points within each group (consecutive points with the same
+ * ID).
```
```suggestion
 * ID). This function can be applied to trajectory data, polygon vertices, linestring vertices, or any grouped point data.
```

trxcllnt commented:
rerun tests

harrism commented:
Ugh, I wish pre-commit would run for github-ui commits.

robertmaynard reviewed (approved):
Aprroving CMake changes

harrism commented:
@gpucibot merge
